### PR TITLE
[config/main.py] Modify reload() function to load configuration from init_cfg.json implicitly

### PR DESCRIPTION
**Pull request by yozhao101**
**- What I did**
I modified the function reload() such that it will load the configuration files config_db.json and init_cfg.json into Config_DB.

**- How I did it**

**- How to verify it**

**- Previous command output (if the output of a command-line utility has changed)**

**- New command output (if the output of a command-line utility has changed)**

### CODE REVIEW

File: config/main.py [hunk 1]
```diff
@@ -506,21 +506,23 @@ def save(filename):
...
-    """Import a previous saved config DB dump file."""
+@click.argument('init_cfg_file', default='/etc/sonic/init_cfg.json', type=click.Path(exists=True))
+def load(filename, init_cfg_file, yes):
+    """Import previous saved config DB dump file and init_cfg.json."""
     if not yes:
-        click.confirm('Load config from the file %s?' % filename, abort=True)
-    command = "{} -j {} --write-to-db".format(SONIC_CFGGEN_PATH, filename)
+        click.confirm('Load config from the files %s and %s?' % (filename, init_cfg_file), abort=True)
+    command = "{} -j {} -j {} --write-to-db".format(SONIC_CFGGEN_PATH, filename, init_cfg_file)
     run_command(command, display_cmd=True)
```
Comment (jleveque) [inline config/main.py]: On second thought, I think we should not load init_cfg.json during `config load`, as this command is meant to load the contents of config_db.json over the existing running config. Since the init_cfg.json should not change until the device is upgraded, there should be no need to load init_cfg.json here. Please undo these changes.

File: config/main.py [hunk 2]
```diff
@@ -506,21 +506,23 @@ def save(filename):
...
 
 @config.command()
 @click.option('-y', '--yes', is_flag=True)
 @click.option('-l', '--load-sysinfo', is_flag=True, help='load system default information (mac, portmap etc) first.')
 @click.argument('filename', default='/etc/sonic/config_db.json', type=click.Path(exists=True))
-def reload(filename, yes, load_sysinfo):
-    """Clear current configuration and import a previous saved config DB dump file."""
+@click.argument('init_cfg_file', default='/etc/sonic/init_cfg.json', type=click.Path(exists=True))
```
Comment (jleveque) [inline config/main.py]: I don't think we need to make this file a user-configurable option. Initial configuration is built into the image and should be fixed. Just make it a constant.

File: config/main.py [hunk 3]
```diff
@@ -544,7 +546,7 @@ def reload(filename, yes, load_sysinfo):
         command = "{} -H -k {} --write-to-db".format(SONIC_CFGGEN_PATH, cfg_hwsku)
         run_command(command, display_cmd=True)
 
-    command = "{} -j {} --write-to-db".format(SONIC_CFGGEN_PATH, filename)
+    command = "{} -j {} -j {} --write-to-db".format(SONIC_CFGGEN_PATH, filename, init_cfg_file)
```
Comment (jleveque) [inline config/main.py]: Here, you should check for the presence of the init_cfg.json file and run the appropriate command based on whether or not it exists. See how it's done in the load_minigraph command on lines 609-612.

File: config/main.py (hunk 1 above)
Comment (yozhao101) [reply config/main.py]: Undo the changes.

File: config/main.py (hunk 2 above)
Comment (yozhao101) [reply config/main.py]: Define a constant string to represent the path of file init_cfg.json and please review it.

File: config/main.py [hunk 4]
```diff
@@ -518,9 +520,9 @@ def load(filename, yes):
 @click.option('-l', '--load-sysinfo', is_flag=True, help='load system default information (mac, portmap etc) first.')
 @click.argument('filename', default='/etc/sonic/config_db.json', type=click.Path(exists=True))
 def reload(filename, yes, load_sysinfo):
-    """Clear current configuration and import a previous saved config DB dump file."""
+    """Clear current configuration and import previous saved config DB dump file and init_cfg.json."""
```
Comment (jleveque) [inline config/main.py]: I don't think it's necessary to change this line to mention init_cfg.json. It may confuse users. It should happen silently. Please change back.

File: config/main.py [hunk 5]
```diff
@@ -518,9 +520,9 @@ def load(filename, yes):
...
     if not yes:
-        click.confirm('Clear current config and reload config from the file %s?' % filename, abort=True)
+        click.confirm('Clear current config and reload config from the files %s and %s?' % (filename, init_cfg_file), abort=True)
```
Comment (jleveque) [inline config/main.py]: I don't think it's necessary to change this line to mention init_cfg.json. It may confuse users. It should happen silently. Please change back.

File: config/main.py (hunk 3 above)
Comment (yozhao101) [reply config/main.py]: Added a if statement to decide whether the file init_cfg.json exists or not.

File: config/main.py [hunk 6]
```diff
@@ -544,6 +546,9 @@ def reload(filename, yes, load_sysinfo):
         command = "{} -H -k {} --write-to-db".format(SONIC_CFGGEN_PATH, cfg_hwsku)
         run_command(command, display_cmd=True)
 
+    if os.path.isfile('/etc/sonic/init_cfg.json'):
+        command = "{} -j {} -j {} --write-to-db".format(SONIC_CFGGEN_PATH, filename, INIT_CFG_FILE)
+
     command = "{} -j {} --write-to-db".format(SONIC_CFGGEN_PATH, filename)
```
Comment (jleveque) [inline config/main.py]: This line should be part of an 'else' clause, Otherwise you're loading 'filename' twice if init_cfg.json exists.

Comment (yozhao101) [reply config/main.py]: My bad and add an else statement.

File: config/main.py (hunk 4 above)
Comment (yozhao101) [reply config/main.py]: Fixed.

File: config/main.py (hunk 5 above)
Comment (yozhao101) [reply config/main.py]: Fixed.

File: config/main.py [hunk 7]
```diff
@@ -544,7 +546,11 @@ def reload(filename, yes, load_sysinfo):
         command = "{} -H -k {} --write-to-db".format(SONIC_CFGGEN_PATH, cfg_hwsku)
         run_command(command, display_cmd=True)
 
-    command = "{} -j {} --write-to-db".format(SONIC_CFGGEN_PATH, filename)
+    if os.path.isfile('/etc/sonic/init_cfg.json'):
+        command = "{} -j {} -j {} --write-to-db".format(SONIC_CFGGEN_PATH, filename, INIT_CFG_FILE)
```
Comment (jleveque) [inline config/main.py]: Have you confirmed this is the appropriate order of the "-j" arguments to first load the init_cfg.json file followed by the config_db.json file? It looks like sonic-cfggen loads the files in the order they are supplied: https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-config-engine/sonic-cfggen#L258

Comment (yozhao101) [reply config/main.py]: For the statement of line 550, sonic-cfggen will first load config_db.json and then init_cfg.json, right?

Comment (jleveque) [reply config/main.py]: That's what it looks like to me. Have you confirmed this? We want to first load init_cf.json, then load config_db.json, overwriting any values which exist in both. Please confirm.

Comment (yozhao101) [reply config/main.py]: I think sonic-cfggen should first load init_cfg.json and then config_db.json.

Comment (jleveque) [reply config/main.py]: If it loads in the order provided, it would first load config_db.json and then init_cfg.json. Please test and verify the behavior and report back.

Comment (jleveque) [reply config/main.py]: You just made a commit to change the order, but you haven't reported back your results. Can you please confirm the order by testing?

Comment (yozhao101) [reply config/main.py]: Joe, I tested on DUT using command **_sonic-cfggen -j /etc/sonic/init_cfg.json -j /etc/sonic/config_db.json --write-to-db_**. In init_cfg.json file, the initial values for "auto_restart" and "high_mem_alert" are "disabled". In config_db.json, the initial values for these two keys are "enabled". After running this command and then using command **_show container feature autorestart_**, the key for "auto_restart" will be overrode to "enabled".

Comment (jleveque) [reply config/main.py]: OK, thanks. Can you also confirm that if you swap the order of the files back to `sonic-cfggen -j /etc/sonic/config_db.json -j /etc/sonic/init_cfg.json --write-to-db` that the results will be wrong, as we assume?

Comment (yozhao101) [reply config/main.py]: Yes, I tested on DUT using the command **_sonic-cfggen -j /etc/sonic/config_db.json -j /etc/sonic/init_cfg.json `--write-to-db_**. Initially the key "auto_restart“ for database in init_cfg.json is set to "disabled" and this key for database is set to "enabled" in config_db.json. After running this command and the command` show container feature autorestart`, the value for auto_restart will be overrode to "disabled".

Comment (jleveque) [reply config/main.py]: Great. Thanks! I believe we saw at least one place in sonic-buildimage where the order is reversed. When you begin work in there, please check for and fix any instances of this.

Comment (yozhao101) [reply config/main.py]: Yes. for sonic-cfggen command, I found the following locations which we can review and may add init_cfg.json. Some locations already load init_cfg.json.  files/image_config/updategraph/updategraph ===> line 10; files/scripts/configdb-load.sh ===> line 10;  files/build_templates/docker_image_ctl.j2 ===> line 111; files/image_config/config-setup/config-setup ===> line 110 and 190;  I will check and change the order.

File: config/main.py [hunk 8]
```diff
@@ -544,7 +546,11 @@ def reload(filename, yes, load_sysinfo):
         command = "{} -H -k {} --write-to-db".format(SONIC_CFGGEN_PATH, cfg_hwsku)
         run_command(command, display_cmd=True)
 
-    command = "{} -j {} --write-to-db".format(SONIC_CFGGEN_PATH, filename)
+    if os.path.isfile('/etc/sonic/init_cfg.json'):
```
Comment (jleveque) [inline config/main.py]: Since you created a constant, please use it here, also:

```
    if os.path.isfile(INIT_CFG_FILE):
```

Comment (yozhao101) [reply config/main.py]: Fixed.
